File: ios/App/App/UnityAdsManager.swift
//
//  UnityAdsManager.swift
//  App
//
//  Unity Ads SDK Manager for iOS
//  Handles initialization, loading, and showing ads
//

import Foundation
import UnityAds

class UnityAdsManager: NSObject, UnityAdsInitializationDelegate, UnityAdsLoadDelegate, UnityAdsShowDelegate {
    
    static let shared = UnityAdsManager()
    
    // Unity Game ID for iOS
    private let gameID = "5968312"
    private let testMode = false // Production mode
    
    // Placement IDs
    private let interstitialPlacement = "Interstitial_iOS"
    private let rewardedPlacement = "Rewarded_iOS"
    
    // Ad ready status
    private var isInterstitialReady = false
    private var isRewardedReady = false
    
    // Reward callback
    var rewardCallback: ((Bool) -> Void)?
    
    // Current view controller (needed for showing ads)
    weak var viewController: UIViewController?
    
    private override init() {
        super.init()
    }
    
    // MARK: - Initialization
    
    func initialize(viewController: UIViewController) {
        self.viewController = viewController
        
        print("[Unity Ads] Initializing with Game ID: \(gameID)")
        UnityAds.initialize(gameID, testMode: testMode, initializationDelegate: self)
    }
    
    // MARK: - UnityAdsInitializationDelegate
    
    func initializationComplete() {
        print("[Unity Ads] ✅ Initialization complete")
        loadInterstitial()
        loadRewarded()
    }
    
    func initializationFailed(_ error: UnityAdsInitializationError, withMessage message: String) {
        print("[Unity Ads] ❌ Initialization failed: \(error) - \(message)")
    }
    
    // MARK: - Load Ads
    
    func loadInterstitial() {
        print("[Unity Ads] Loading interstitial...")
        UnityAds.load(interstitialPlacement, loadDelegate: self)
    }
    
    func loadRewarded() {
        print("[Unity Ads] Loading rewarded...")
        UnityAds.load(rewardedPlacement, loadDelegate: self)
    }
    
    // MARK: - Show Ads
    
    func showInterstitial() {
        guard isInterstitialReady, let vc = viewController else {
            print("[Unity Ads] ⚠️ Interstitial not ready or no view controller")
            return
        }
        
        print("[Unity Ads] Showing interstitial...")
        UnityAds.show(vc, placementId: interstitialPlacement, showDelegate: self)
    }
    
    func showRewarded(completion: @escaping (Bool) -> Void) {
        guard isRewardedReady, let vc = viewController else {
            print("[Unity Ads] ⚠️ Rewarded not ready or no view controller")
            completion(false)
            return
        }
        
        self.rewardCallback = completion
        print("[Unity Ads] Showing rewarded...")
        UnityAds.show(vc, placementId: rewardedPlacement, showDelegate: self)
    }
    
    // MARK: - UnityAdsLoadDelegate
    
    func unityAdsAdLoaded(_ placementId: String) {
        print("[Unity Ads] ✅ Ad loaded: \(placementId)")
        
        if placementId == interstitialPlacement {
            isInterstitialReady = true
        } else if placementId == rewardedPlacement {
            isRewardedReady = true
        }
    }
    
    func unityAdsAdFailed(toLoad placementId: String, withError error: UnityAdsLoadError, withMessage message: String) {
        print("[Unity Ads] ❌ Ad failed to load: \(placementId) - \(error) - \(message)")
        
        if placementId == interstitialPlacement {
            isInterstitialReady = false
        } else if placementId == rewardedPlacement {
            isRewardedReady = false
        }
    }
    
    // MARK: - UnityAdsShowDelegate
    
    func unityAdsShowComplete(_ placementId: String, withFinish state: UnityAdsShowCompletionState) {
        print("[Unity Ads] ✅ Ad show complete: \(placementId) - \(state.rawValue)")
        
        if placementId == interstitialPlacement {
            isInterstitialReady = false
            // Reload after 3 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 3.0) {
                self.loadInterstitial()
            }
        } else if placementId == rewardedPlacement {
            isRewardedReady = false
            
            // Check if user completed the ad
            if state == .showCompletionStateCompleted {
                rewardCallback?(true)
            } else {
                rewardCallback?(false)
            }
            rewardCallback = nil
            
            // Reload after 3 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 3.0) {
                self.loadRewarded()
            }
        }
    }
    
    func unityAdsShowFailed(_ placementId: String, withError error: UnityAdsShowError, withMessage message: String) {
        print("[Unity Ads] ❌ Ad show failed: \(placementId) - \(error) - \(message)")
        
        if placementId == interstitialPlacement {
            isInterstitialReady = false
            loadInterstitial()
        } else if placementId == rewardedPlacement {
            isRewardedReady = false
            rewardCallback?(false)
            rewardCallback = nil
            loadRewarded()
        }
    }
    
    func unityAdsShowStart(_ placementId: String) {
        print("[Unity Ads] 📺 Ad show start: \(placementId)")
    }
    
    func unityAdsShowClick(_ placementId: String) {
        print("[Unity Ads] 👆 Ad clicked: \(placementId)")
    }
    
    // MARK: - Status Check
    
    func isInterstitialAdReady() -> Bool {
        return isInterstitialReady
    }
    
    func isRewardedAdReady() -> Bool {
        return isRewardedReady
    }
}

File: android/app/src/main/java/com/bilgoo/quizapp/UnityAdsManager.java
package com.bilgoo.quizapp;

import android.app.Activity;
import android.util.Log;
import com.unity3d.ads.IUnityAdsInitializationListener;
import com.unity3d.ads.IUnityAdsLoadListener;
import com.unity3d.ads.IUnityAdsShowListener;
import com.unity3d.ads.UnityAds;
import com.unity3d.ads.UnityAdsShowOptions;

public class UnityAdsManager {
    private static final String TAG = "UnityAdsManager";
    private static final String GAME_ID = "5968313"; // Android Game ID
    private static final boolean TEST_MODE = false; // Production mode
    
    private static final String INTERSTITIAL_PLACEMENT = "Interstitial_Android"; // Unity default placement
    private static final String REWARDED_PLACEMENT = "Rewarded_Android"; // Unity default placement
    
    private Activity activity;
    private boolean isInitialized = false;
    private boolean isInterstitialReady = false;
    private boolean isRewardedReady = false;
    
    public interface RewardCallback {
        void onRewarded();
        void onRewardFailed();
    }
    
    public UnityAdsManager(Activity activity) {
        this.activity = activity;
        initializeAds();
    }
    
    private void initializeAds() {
        Log.d(TAG, "Initializing Unity Ads with Game ID: " + GAME_ID);
        
        UnityAds.initialize(activity.getApplicationContext(), GAME_ID, TEST_MODE, new IUnityAdsInitializationListener() {
            @Override
            public void onInitializationComplete() {
                Log.d(TAG, "Unity Ads initialization complete");
                isInitialized = true;
                loadInterstitial();
                loadRewarded();
            }
            
            @Override
            public void onInitializationFailed(UnityAds.UnityAdsInitializationError error, String message) {
                Log.e(TAG, "Unity Ads initialization failed: " + error + " - " + message);
                isInitialized = false;
            }
        });
    }
    
    public void loadInterstitial() {
        if (!isInitialized) return;
        
        UnityAds.load(INTERSTITIAL_PLACEMENT, new IUnityAdsLoadListener() {
            @Override
            public void onUnityAdsAdLoaded(String placementId) {
                Log.d(TAG, "Interstitial ad loaded: " + placementId);
                isInterstitialReady = true;
            }
            
            @Override
            public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
                Log.e(TAG, "Interstitial failed to load: " + error + " - " + message);
                isInterstitialReady = false;
            }
        });
    }
    
    public void showInterstitial() {
        if (!isInitialized || !isInterstitialReady) {
            Log.w(TAG, "Interstitial not ready");
            return;
        }
        
        UnityAds.show(activity, INTERSTITIAL_PLACEMENT, new UnityAdsShowOptions(), new IUnityAdsShowListener() {
            @Override
            public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
                Log.e(TAG, "Interstitial show failed: " + error + " - " + message);
                isInterstitialReady = false;
                loadInterstitial();
            }
            
            @Override
            public void onUnityAdsShowStart(String placementId) {
                Log.d(TAG, "Interstitial show start: " + placementId);
            }
            
            @Override
            public void onUnityAdsShowClick(String placementId) {
                Log.d(TAG, "Interstitial clicked: " + placementId);
            }
            
            @Override
            public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
                Log.d(TAG, "Interstitial show complete: " + state);
                isInterstitialReady = false;
                loadInterstitial();
            }
        });
    }
    
    public void loadRewarded() {
        if (!isInitialized) return;
        
        UnityAds.load(REWARDED_PLACEMENT, new IUnityAdsLoadListener() {
            @Override
            public void onUnityAdsAdLoaded(String placementId) {
                Log.d(TAG, "Rewarded ad loaded: " + placementId);
                isRewardedReady = true;
            }
            
            @Override
            public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
                Log.e(TAG, "Rewarded failed to load: " + error + " - " + message);
                isRewardedReady = false;
            }
        });
    }
    
    public void showRewarded(final RewardCallback callback) {
        if (!isInitialized || !isRewardedReady) {
            Log.w(TAG, "Rewarded ad not ready");
            if (callback != null) callback.onRewardFailed();
            return;
        }
        
        UnityAds.show(activity, REWARDED_PLACEMENT, new UnityAdsShowOptions(), new IUnityAdsShowListener() {
            @Override
            public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
                Log.e(TAG, "Rewarded show failed: " + error + " - " + message);
                isRewardedReady = false;
                loadRewarded();
                if (callback != null) callback.onRewardFailed();
            }
            
            @Override
            public void onUnityAdsShowStart(String placementId) {
                Log.d(TAG, "Rewarded show start: " + placementId);
            }
            
            @Override
            public void onUnityAdsShowClick(String placementId) {
                Log.d(TAG, "Rewarded clicked: " + placementId);
            }
            
            @Override
            public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
                Log.d(TAG, "Rewarded show complete: " + state);
                isRewardedReady = false;
                loadRewarded();
                
                if (state == UnityAds.UnityAdsShowCompletionState.COMPLETED) {
                    if (callback != null) callback.onRewarded();
                } else {
                    if (callback != null) callback.onRewardFailed();
                }
            }
        });
    }
    
    public boolean isInterstitialReady() {
        return isInterstitialReady;
    }
    
    public boolean isRewardedReady() {
        return isRewardedReady;
    }
}
